pysrc/bytewax/connectors/kafka/operators.py

```diff
@@ -0,0 +1,309 @@
+        topic:
+            Topic to produce to.
+        add_config:
+            Any additional configuration properties. See the `rdkafka`
+            [docs](https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md)
+            for options.
+    """
+    return op.output("kafka_output", up, KafkaSink(brokers, topic, add_config))
+
+
+@operator
+def deserialize_key(
+    step_id: str,
+    up: Stream[KafkaMessage[KafkaKey, V]],
+    deserializer: SchemaDeserializer,
+) -> KafkaStreams[Dict, V, KafkaKey, V]:
+    """Deserialize the key of a KafkaMessage using the provided SchemaDeserializer.
+
+    Returns an object with two attributes:
+        - .oks: A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+        - .errors: A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    """
+
+    def shim_mapper(msg: KafkaMessage) -> KafkaMessage:
+        try:
+            key = deserializer.de(msg.key)
+            # _kafka_error_split casts the type of the `oks` stream
+            # to a KafkaMessage with types inferred from the return type
+            # of this function. The casts doesn't check anything though,
+            # so we need to make sure we are telling the truth:
+            assert isinstance(key, dict)
+            return msg.with_key(key)
+        except Exception as e:
+            return msg.with_error(KafkaError(KafkaError._KEY_DESERIALIZATION, f"{e}"))
+
+    return op.map("map", up, shim_mapper).then(_kafka_error_split, "split")
+
+
+@operator
+def deserialize_value(
+    step_id: str,
+    up: Stream[KafkaMessage[K, KafkaValue]],
+    deserializer: SchemaDeserializer,
+) -> KafkaStreams[K, Dict, K, KafkaValue]:
+    """Deserialize the value of a KafkaMessage using the provided SchemaDeserializer.
+
+    Returns an object with two attributes:
+        - .oks: A stream of `KafkaMessage`s where `KafkaMessage.error is None`
+        - .errors: A stream of `KafkaMessage`s where `KafkaMessage.error is not None`
+    """
+
+    def shim_mapper(msg: KafkaMessage) -> KafkaMessage:
+        try:
+            value = deserializer.de(msg.value)
+            # _kafka_error_split casts the type of the `oks` stream
+            # to a KafkaMessage with types inferred from the return type
+            # of this function. The casts doesn't check anything though,
+            # so we need to make sure we are telling the truth:
+            assert isinstance(value, dict)
+            return msg.with_value(value)
+        except Exception as e:
+            return msg.with_error(KafkaError(KafkaError._VALUE_DESERIALIZATION, f"{e}"))
+
+    return op.map("map", up, shim_mapper).then(_kafka_error_split, "split_err")
+
+
+@operator
+def deserialize(
+    step_id: str,
+    up: Stream[KafkaMessage[KafkaKey, KafkaValue]],
+    *,
+    key_deserializer: SchemaDeserializer,
+    val_deserializer: SchemaDeserializer,
+) -> KafkaStreams[Dict, Dict, KafkaKey, KafkaValue]:
```

Comment:
I think this should be `KafkaStreams[Dict, Dict, Union[KafkaKey, Dict], Union[KafkaValue, Dict]]` since we know the previous and expected types, and the messages with an error should either be the unprocessed type, or the processed one. I fixed some type hints around, and now the error streams seems to be properly typed